docker/local-development.md

```diff
@@ -0,0 +1,189 @@
+# Docker for Local Development
+
+Docker for local development was piloted on `dedupe-service`, a repository to
+which some DataMade employees do not have access. Luckily, we can abstract general
+patterns for Dockerizing an application to guide our next effort.
+
+## Overview
+
+A containerized local development environment has four components:
+
+1. [A Dockerfile](#dockerfile), containing building instructions for the application itself.
+2. [A database initialization script](#scriptsinit-dbsh)
+that creates your database and installs any extensions
+2. [A root `docker-compose.yml` file](#docker-composeyml)
+that declares the application and its dependent services
+3. [A `tests/docker-compose.yml` file](#testsdocker-composeyml)
+that overrides the application service in the root file, in order to run the tests
+
+### 1. `Dockerfile`
+
+```Dockerfile
+# Extend the base Python image
+# See https://hub.docker.com/_/python for version options
+# N.b., there are many options for Python images. We used the plain
+# version number in the pilot. YMMV. See this post for a discussion of
+# some options and their pros and cons:
+# https://pythonspeed.com/articles/base-image-python-docker-images/
+FROM python:<PYTHON_VERSION>
+
+# Give ourselves some credit
+LABEL maintainer "DataMade <info@datamade.us>"
+
+# Install and upgrade pip
+# This may not be necessary, depending on your base Python image
+RUN apt-get update
+RUN apt-get install -y python-pip
+RUN pip install --upgrade pip setuptools
+
+# Inside the container, create an app directory and switch into it
+RUN mkdir /app
+WORKDIR /app
+
+# Copy the requirements file into the app directory, and install them. Copy
+# only the requirements file, so Docker can cache this build step. Otherwise,
+# the requirements must be reinstalled every time you build the image after
+# the app code changes. See this post for further discussion of strategies
+# for building lean and efficient containers:
+# https://blog.realkinetic.com/building-minimal-docker-containers-for-python-applications-37d0272c52f3
+COPY ./requirements.txt /app/requirements.txt
+RUN pip install --no-cache-dir -r requirements.txt
+
+# Copy the contents of the current host directory (i.e., our app code) into
+# the container.
+COPY . /app
+```
```

Comment:
These example scripts are great, and make me think that a containerized Django app template would be helpful. Maybe a next step to take after the pilot project?

Reply:
yes, absolutely!